Firestore/core/src/firebase/firestore/util/string_win.cc

```diff
@@ -0,0 +1,125 @@
+  // terminated so we must pass an explicit length to MultiByteToWideChar. This
+  // means the result is the number of wchar_t required to hold the result,
+  // excluding the null terminator.
+  int output_len =
+      ::MultiByteToWideChar(CP_UTF8, 0, input.data(), input_len, nullptr, 0);
+  if (output_len == 0) {
+    DWORD error = GetLastError();
+    HARD_FAIL("Utf8ToNative failed with code %s: %s", error,
+              LastErrorMessage(error));
+  }
+  HARD_ASSERT(output_len > 0);
+
+  int output_terminated_len = output_len + 1;
+  std::wstring output(output_terminated_len, L'\0');
+  int result = ::MultiByteToWideChar(CP_UTF8, 0, input.data(), input_len,
+                                     &output[0], output_len);
+  HARD_ASSERT(result > 0 && result < output_terminated_len &&
+              output[output_len] == '\0');
+
+  output.resize(result);
+  return output;
+}
+
+std::string NativeToUtf8(const std::wstring& input) {
+  return NativeToUtf8(input.c_str(), input.size());
+}
+
+std::string NativeToUtf8(const wchar_t* input, size_t input_size) {
+  int input_len = static_cast<int>(input_size);
+  if (input_len == 0) {
+    return "";
+  }
+
+  // The input buffer may contain embedded nulls and isn't necessarily null
+  // terminated so we must pass an explicit length to WideCharToMultiByte. This
+  // means the result is the number of char required to hold the result,
+  // excluding the null terminator.
+  int output_len = ::WideCharToMultiByte(CP_UTF8, 0, input, input_len, nullptr,
+                                         0, nullptr, nullptr);
+  if (output_len == 0) {
+    DWORD error = GetLastError();
+    HARD_FAIL("NativeToUtf8 failed with code %s: %s", error,
+              LastErrorMessage(error));
+  }
+  HARD_ASSERT(output_len > 0);
+
+  int output_terminated_len = output_len + 1;
+  std::string output(output_terminated_len, '\0');
+  int result = ::WideCharToMultiByte(CP_UTF8, 0, input, input_len, &output[0],
+                                     output_len, nullptr, nullptr);
+  HARD_ASSERT(result > 0 && result < output_terminated_len &&
+              output[output_len] == '\0');
+
+  output.resize(result);
+  return output;
+}
+
+std::string LastErrorMessage(DWORD last_error) {
+  // output_len excludes the trailing null.
+  wchar_t* error_text = nullptr;
+  DWORD output_len = FormatMessageW(
+      FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_ALLOCATE_BUFFER |
+          FORMAT_MESSAGE_IGNORE_INSERTS,
+      nullptr, last_error, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+      (wchar_t*)&error_text, 0, nullptr);
+  std::string formatted;
+  if (output_len == 0 || error_text == nullptr) {
+    formatted = util::StringFormat(
+        "error %s; unknown error %s while getting error text", last_error,
+        output_len);
+  } else {
+    formatted = NativeToUtf8(error_text, output_len);
+    if (absl::EndsWith(formatted, "\r\n")) {
+      formatted.resize(formatted.size() - 2);
+    }
+  }
+
+  LocalFree(error_text);
```

Comment:
This function isn't exception safe.

OTOH, we don't use them, there's no opportunity for user code to be called here, and I don't think any of the MS functions that you've used use them either. Maybe an allocation exception could occur? Worst case: we don't free memory during an error event.

So I don't think this matters.

However, if you want to "fix" this, you could just pre-allocate 64k and pass in that buffer. Why 64k?
a) that's the maximum buffer size that FormatMessageW accepts, so you can't pass in a bigger one
b) it's apparently what FormatMessageW does anyways if you pass FORMAT_MESSAGE_ALLOCATE_BUFFER. At least, that's my reading of https://blogs.msdn.microsoft.com/oldnewthing/20120210-00/?p=8333: "It so happens that in Windows Vista, the fancy buffer management technique was abandoned, and the code just allocates 64KB of memory up front" But I haven't read through that in detail, so I could be wrong.

So I don't think you're saving any memory by letting the function alloc for you.

Reply:
I wouldn't say that I was intending to save memory, exactly. These functions from the dawn of Win32 are sufficiently byzantine that it's worth following somebody [else's example](https://stackoverflow.com/questions/455434/how-should-i-use-formatmessage-properly-in-c). Indeed one day somebody's going to figure out how to plug stackoverflow directly into the editor and I will be obsolete.

I've changed this to pre-allocate a buffer. Interesting to note that at least on Windows 10, `FormatMessageW` seemed to readily accept any size buffer.